### PR TITLE
Enforcing `_evaluate_t` and `_evaluate` signatures

### LINKED ISSUE #678: `_evaluate` and `_evaluate_t` should only deal with arrays
#618 enforced the high level interface for these basis methods and closed #553 

> So the idea here is that:
> 
> - All `evaluate`s should take an `ndarray` as input and output a `Volume`/`Image`
> - All `evaluate_t`s and `expand`s should take an `Volume`/`Image` as input (but will cast an `ndarray` to the appropriate cast after logging the warning) and redurn an `ndarray`
> 
> Certain parts of this is implemented in certain `Basis` classes and other parts in other classes. To sort this out, the best way may be to define `evaluate`, `evaluate_t`, and `expand` in a superclass and have these do the necessary checks and casts, then call `_evaluate`, `_evaluate_t`, `_expand`, and so on that will be overriden in subclasses and operate on `ndarrays`. Obviously this would have to be somewhat smart in checking whether the basis is 2D (in which case we're dealing with `Image`s) or 3D (in which case we're dealing with `Volume`s).
> 
> _Originally posted by @janden in https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/553#issuecomment-1051030251_
> 
> 

However, we want the private methods to also have a common interface dealing with only arrays, with casting done at the `Basis` public method level. This may involve digging deep and tweaking the various basis-specific methods. Dealing with `mat_evaluate` and `mat_evaluate_t` may also be an issue.

### PR DESCRIPTION
**Pull request by chris-langfield**
#678 

`Basis` public methods `evaluate` and `evaluate_t` should accept a numpy array and return an `Image/Volume`, and vice versa. But the private methods should only deal with numpy arrays. This PR enforces that change and cleans up the `Basis` class.

### CODE REVIEW

File: src/aspire/basis/basis.py [hunk 1]
```diff
@@ -110,16 +113,13 @@ def evaluate_t(self, v):
                 f" Inconsistent dtypes v: {v.dtype} self: {self.dtype}"
             )
 
-        if not isinstance(v, Image) and not isinstance(v, Volume):
-            if self.ndim == 2:
-                _class = Image
-            elif self.ndim == 3:
-                _class = Volume
+        if not isinstance(v, self.cls):
             logger.warning(
```
Comment (garrettwrong) [inline src/aspire/basis/basis.py:117]: I'm considering whether we should start to use more python warnings. What do you think?

**Review by garrettwrong — DISMISSED** [review]
I realize now some people might prefer `self.cls` be `self._cls`. I don't mind either way, buf if you are one of those people, free to change it :).

Just the one comment question, otherwise looks good. Glad it worked out.

Thanks!

**Comment by chris-langfield** [comment]
Actually, I do agree with that. Unless we were to rename it to something more descriptive such as `data_cls` or `shape_cls` and expose it (I can't think of a use case for that), then I would make it private.

File: src/aspire/basis/basis.py (hunk 1 above)
Comment (chris-langfield) [reply src/aspire/basis/basis.py:117]: I agree this is a scenario where `warning.warn` makes more sense. Should we make an issue so it's out there that we want to change these as we come across them?

File: src/aspire/basis/ffb_2d.py [hunk 2]
```diff
@@ -197,11 +197,10 @@ def _evaluate_t(self, x):
         freqs = np.reshape(self._precomp["freqs"], (2, n_r * n_theta))
 
         # number of 2D image samples
-        n_images = x.n_images
-        x_data = x.data
+        n_images = x.shape[0]
```
Comment (j-c-c) [inline src/aspire/basis/ffb_2d.py:200]: I think the docstring `:return:` should say "The evaluation of the coefficient array `x`..."

File: src/aspire/basis/ffb_3d.py [hunk 3]
```diff
@@ -289,8 +289,6 @@ def _evaluate_t(self, x):
             `self.count` and whose remaining dimensions correspond to higher
```
Comment (j-c-c) [inline src/aspire/basis/ffb_3d.py:289]: Same as above: `v` --> `x` in docstring `return`

File: src/aspire/basis/polar_2d.py [hunk 4]
```diff
@@ -107,11 +107,11 @@ def _evaluate_t(self, x):
             Fourier grid. This is an array of vectors whose first dimension
             corresponds to x.n_images, and last dimension equals `self.count`.
         """
-        nimgs = x.n_images
+        nimgs = x.shape[0]
```
Comment (j-c-c) [inline src/aspire/basis/polar_2d.py:110]: Same as above: `v` --> `x` in docstring

**Review by j-c-c — COMMENTED** [review]
Looks good. Just a couple small docstring changes.

File: src/aspire/basis/polar_2d.py (hunk 4 above)
Comment (chris-langfield) [reply src/aspire/basis/polar_2d.py:110]: shoot thanks for catching these. Fixed

**Pull request merged**